### PR TITLE
feat(*): port several PDK and shared llm module from ai-proxy-advanced feature

### PR DESCRIPTION
**Pull request by fffonion**
### Summary

This PR adds several PDK methods to allow ai-proxy-advanced plugin to do customized load balancing and failover.
It also moves the ai-proxy handler to a shared module so that both ai-proxy{,-advanced} plugins can reuse same codebase.

https://github.com/Kong/kong-ee/pull/9562

### Checklist

- [x] The Pull Request has tests
- [na] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [na] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

AG-12

### CODE REVIEW

File: kong/runloop/balancer/init.lua [hunk 1]
```diff
@@ -359,6 +359,16 @@ local function execute(balancer_data, ctx)
     balancer_data.balancer_handle = handle
 
   else
+    -- Note: balancer_data.retry_callback is only set by PDK once in access phase
```
Comment (liverpool8056) [inline kong/runloop/balancer/init.lua:362]: It seems that the `retry_callback` can be invoked only when `balancer == nil`, is it expected to work like this?

In addition, should we wrap retry_callback with pcall?

Comment (fffonion) [reply kong/runloop/balancer/init.lua:362]: > It seems that the retry_callback can be invoked only when balancer == nil, is it expected to work like this?

Yes, if we skip the balancer lookup by doing `kong.service.set_target`,  i.e. currently request doesn't have a upstream, then `balancer, upstream = balancers.get_balancer(balancer_data)` will return no balancer. Actually, this else branch is just made to support `kong.service.set_target`.
This matches our expectation, retry_callback will not be called if Kong's balancer is not skipped.

> In addition, should we wrap retry_callback with pcall?

yes, adding that now.

Comment (fffonion) [reply kong/runloop/balancer/init.lua:362]: I will address all comments in EE side of PR and port here

**Comment by fffonion** [comment]
Reviewers: please review https://github.com/Kong/kong-ee/pull/9562 first for a whole picture, I will then address comments there and port here.

File: kong/runloop/balancer/init.lua (hunk 1 above)
Comment (fffonion) [reply kong/runloop/balancer/init.lua:362]: fixed

**Review by jschmid1 — APPROVED** [review]

**Review by oowl — APPROVED** [review]

**Comment by fffonion** [comment]
ce-ee cherrypick not needed, it's part of the ai-proxy-advanced PR and is merged

**Pull request merged**